### PR TITLE
Feature/rework key manager shm

### PR DESCRIPTION
**Pull request by Alexey-N-Chernyshov**
### Description of the Change
Rework key manager with shared_model.

### Benefits
One another step to the bright future of shared_model.

### Possible Drawbacks 
no

### CODE REVIEW

File: iroha-cli/main.cpp [hunk 1]
```diff
@@ -158,20 +156,17 @@ int main(int argc, char *argv[]) {
...
-      keypair = manager.loadKeys();
-    }
+    auto keypair = FLAGS_pass_phrase.size() != 0
+        ? manager.loadKeys(FLAGS_pass_phrase)
+        : manager.loadKeys();
     if (not keypair) {
       logger->error(
           "Cannot load specified keypair, or keypair is invalid. Path: {}, "
-          "keypair name: {}. Use --key_path to path to your keypair. \nMaybe wrong pass phrase (\"{}\")?",
+          "keypair name: {}. Use --key_path to path to your keypair. \n"
```
Comment (l4l) [inline iroha-cli/main.cpp]: fix the typo as well plz: `--key_path to path to your keypair` -> `--key_path with path of your keypair`

File: irohad/consensus/yac/impl/yac_crypto_provider_impl.hpp [hunk 2]
```diff
@@ -19,13 +19,15 @@
 #define IROHA_YAC_CRYPTO_PROVIDER_IMPL_HPP
 
 #include "consensus/yac/yac_crypto_provider.hpp"
+#include "cryptography/keypair.hpp"
```
Comment (l4l) [inline irohad/consensus/yac/impl/yac_crypto_provider_impl.hpp]: Forward decl seems enough here

Comment (l4l) [reply irohad/consensus/yac/impl/yac_crypto_provider_impl.hpp]: check the rest as well plz

File: libs/crypto/keys_manager_impl.cpp [hunk 3]
```diff
@@ -17,37 +17,15 @@
...
-#include <algorithm>
 #include <fstream>
-#include <random>
-#include <utility>
 
 #include "common/byteutils.hpp"
-#include "common/types.hpp"  // for keypair_t, pubkey_t, privkey_t
-#include "cryptography/ed25519_sha3_impl/internal/ed25519_impl.hpp"
+#include "cryptography/crypto_provider/crypto_defaults.hpp"
 #include "cryptography/ed25519_sha3_impl/internal/sha3_hash.hpp"
```
Comment (l4l) [inline libs/crypto/keys_manager_impl.cpp]: Remove that header as well plz. Hash function can be replaced with custom blob

File: libs/crypto/keys_manager_impl.cpp [hunk 4]
```diff
@@ -69,39 +47,25 @@ namespace iroha {
...
+    try {
+      auto test = shared_model::crypto::Blob(sha3_256("12345").to_string());
+      auto sig =
+          shared_model::crypto::DefaultCryptoAlgorithmType::sign(test, keypair);
+      if (not shared_model::crypto::DefaultCryptoAlgorithmType::verify(
+              sig, test, keypair.publicKey())) {
+        log_->error("key validation failed");
+        return false;
+      }
+    } catch (const BadFormatException &exception) {
```
Comment (l4l) [inline libs/crypto/keys_manager_impl.cpp]: Where is the exception may be throwed?

File: libs/crypto/keys_manager_impl.cpp [hunk 5]
```diff
@@ -118,59 +82,66 @@ namespace iroha {
...
   }
 
-  boost::optional<keypair_t> KeysManagerImpl::loadKeys() {
+  boost::optional<shared_model::crypto::Keypair> KeysManagerImpl::loadKeys() {
     std::string pub_key;
     std::string priv_key;
 
-    if (not loadFile(account_name_ + kPubExt, pub_key)
-        or not loadFile(account_name_ + kPrivExt, priv_key))
+    if (not loadFile(account_name_ + public_key_extension, pub_key)
```
Comment (l4l) [inline libs/crypto/keys_manager_impl.cpp]: Revert that plz, constants should follow [our style guide](https://google.github.io/styleguide/cppguide.html#Constant_Names)
Also prefer not changing things, until that is necessary

File: libs/crypto/keys_manager_impl.cpp [hunk 6]
```diff
@@ -69,39 +47,25 @@ namespace iroha {
...
         log_(logger::log("KeysManagerImpl")) {}
 
-  bool KeysManagerImpl::validate(const keypair_t &keypair) const {
-    std::string test = "12345";
-    auto sig =
-        sign(sha3_256(test).to_string(), keypair.pubkey, keypair.privkey);
-    if (not verify(sha3_256(test).to_string(), keypair.pubkey, sig)) {
-      log_->error("key validation failed");
+  bool KeysManagerImpl::validate(
+      const shared_model::crypto::Keypair &keypair) const {
```
Comment (l4l) [inline libs/crypto/keys_manager_impl.cpp]: I think `using namespace shared_model::crypto` may significantly simplify code of that file

File: irohad/consensus/yac/impl/yac_crypto_provider_impl.hpp (hunk 2 above)
Comment (Alexey-N-Chernyshov) [reply irohad/consensus/yac/impl/yac_crypto_provider_impl.hpp]: no, there is member initialization

File: libs/crypto/keys_manager_impl.cpp (hunk 4 above)
Comment (Alexey-N-Chernyshov) [reply libs/crypto/keys_manager_impl.cpp]: Here https://github.com/hyperledger/iroha/blob/9e0e19035308c6ebaf706f709c5b7b3ac46e708b/libs/common/types.hpp#L110
from https://github.com/hyperledger/iroha/blob/9e0e19035308c6ebaf706f709c5b7b3ac46e708b/shared_model/cryptography/ed25519_sha3_impl/signer.cpp#L29

File: irohad/consensus/yac/impl/yac_crypto_provider_impl.hpp (hunk 2 above)
Comment (l4l) [reply irohad/consensus/yac/impl/yac_crypto_provider_impl.hpp]: Ehm, may you point that place out?

File: libs/crypto/keys_manager_impl.cpp (hunk 4 above)
Comment (l4l) [reply libs/crypto/keys_manager_impl.cpp]: Nice catch! Nevertheless it seems redundant after your fixes

File: irohad/consensus/yac/impl/yac_crypto_provider_impl.hpp (hunk 2 above)
Comment (Alexey-N-Chernyshov) [reply irohad/consensus/yac/impl/yac_crypto_provider_impl.hpp]: https://github.com/hyperledger/iroha/pull/1185/files/7749b45e4bae3ef8e494d9f0949842acc0f7b00f#diff-bfaf89f725240ddf0dbe056db3ef5183R41

File: libs/crypto/keys_manager_impl.cpp (hunk 4 above)
Comment (Alexey-N-Chernyshov) [reply libs/crypto/keys_manager_impl.cpp]: Yes, but i don't want exception to be uncaught.

File: irohad/consensus/yac/impl/yac_crypto_provider_impl.hpp (hunk 2 above)
Comment (Alexey-N-Chernyshov) [reply irohad/consensus/yac/impl/yac_crypto_provider_impl.hpp]: *Instantination, not initialization

Comment (l4l) [reply irohad/consensus/yac/impl/yac_crypto_provider_impl.hpp]: Seems like a bug, ok then

File: libs/crypto/keys_manager.hpp [hunk 7]
```diff
@@ -34,31 +42,31 @@ namespace iroha {
...
-    virtual boost::optional<iroha::keypair_t> loadKeys() = 0;
-
     /**
      * Create keys a new keypair and store it encrypted on disk
      * @param pass_phrase is a password for the keys
      * @return false if create account failed
      */
     virtual bool createKeys(const std::string &pass_phrase) = 0;
 
+    /**
```
Comment (x3medima17) [inline libs/crypto/keys_manager.hpp]: Why did you change the order of functions?

Comment (Alexey-N-Chernyshov) [reply libs/crypto/keys_manager.hpp]: It seems more logical to have overloaded function in one place.
